wincode-derive/src/schema_read.rs

```diff
@@ -225,27 +230,232 @@ fn impl_struct_extensions(args: &SchemaArgs) -> Result<TokenStream> {
+        let where_clause = builder_generics.where_clause.as_ref();
+        quote! {
+            /// A helper struct that provides convenience methods for reading and writing to a `MaybeUninit` struct
+            /// with a bit-set tracking the initialization state of the fields.
+            ///
+            /// The builder will drop all initialized fields in reverse order on drop. When the struct is fully initialized,
+            /// you **must** call `assume_init_forget` or `assume_init_mut` to forget the builder. Otherwise, all the
+            /// initialized fields will be dropped when the builder is dropped.
+            #[must_use]
+            #vis struct #builder_ident < #(#generic_lifetimes,)* #(#generic_const,)* #(#generic_type_params,)* > #where_clause {
+                inner: &'_wincode_inner mut core::mem::MaybeUninit<#builder_dst>,
+                init_set: #builder_bit_set_ty,
+            }
+        }
+    };
+
+    let builder_drop_impl = {
+        // Drop all initialized fields in reverse order.
+        let drops = fields.iter().rev().enumerate().map(|(index, field)| {
+            // Compute the actual index relative to the reversed iterator.
+            let real_index = fields.len() - 1 - index;
+            let field_ident = field.struct_member_ident(real_index);
+            // The corresponding bit for the field.
+            let bit_set_index = LitInt::new(&(1 << real_index).to_string(), Span::call_site());
+            quote! {
+                if self.init_set & #bit_set_index != 0 {
+                    // SAFETY: We are dropping an initialized field.
+                    unsafe {
+                        ptr::drop_in_place(&raw mut (*dst_ptr).#field_ident);
+                    }
+                }
+            }
+        });
+        quote! {
+            impl #builder_impl_generics Drop for #builder_ident #builder_ty_generics #builder_where_clause {
+                fn drop(&mut self) {
+                    let dst_ptr = self.inner.as_mut_ptr();
+                    #(#drops)*
+                }
+            }
+        }
+    };
+
+    let builder_impl = quote! {
+        impl #builder_impl_generics #builder_ident #builder_ty_generics #builder_where_clause {
+            #vis const fn from_maybe_uninit_mut(inner: &'_wincode_inner mut MaybeUninit<#builder_dst>) -> Self {
+                Self {
+                    inner,
+                    init_set: 0,
+                }
+            }
+
+            /// Assume the builder is fully initialized, and return a mutable reference to the inner `MaybeUninit` struct.
+            ///
+            /// The builder will be forgotten, so the drop logic will not longer run.
+            ///
+            /// # Safety
+            ///
+            /// Calling this when the content is not yet fully initialized causes undefined behavior: it is up to the caller
+            /// to guarantee that the `MaybeUninit<T>` really is in an initialized state.
```

Comment:
maybe we could assert that all `init_set` bits are set...? I guess this could be over-restrictive, i.e. if user passed in `MaybeUninit` that was already partially initialized and won't like the builder to require calling assume init for those fields, but it seems worth considering